### PR TITLE
Assert WebSession is not null

### LINKED ISSUE #14975: WebSessionServerOAuth2AuthorizedClientRepository throws NPE when session is null
**Describe the bug**
[This line](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java#L70) `session.getAttributes() `throws NPE, if we have WebSession bean configuration like this to disable WebSession:

``` 
@Bean
    fun webSessionManager(): WebSessionManager {
        // Emulate SessionCreationPolicy.STATELESS
        return WebSessionManager { exchange: ServerWebExchange? -> Mono.empty() }
    }
 ``` 

**Expected behavior**
I thing we may do null-safety when we are trying to call getAuthorizedClients in [this place](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java#L99) with nullable session.

**Configuration**
Spring Cloud Gateway.

```
@Bean
    fun authorizedClientManager(
        clientRegistrationRepository: ReactiveClientRegistrationRepository,
        authorizedClientRepository: ServerOAuth2AuthorizedClientRepository
    ): ReactiveOAuth2AuthorizedClientManager {
        val authorizedClientProvider = ReactiveOAuth2AuthorizedClientProviderBuilder
            .builder()
            .clientCredentials()
            .build()
        val authorizedClientManager = DefaultReactiveOAuth2AuthorizedClientManager(
            clientRegistrationRepository, authorizedClientRepository
        )
        authorizedClientManager.setAuthorizedClientProvider(authorizedClientProvider)
        return authorizedClientManager
    }

    @Bean
    fun springSecurityFilterChainNonProd(
        httpSecurity: ServerHttpSecurity,
        someIntrospector: SomeIntrospector,
    ): SecurityWebFilterChain? {
        return httpSecurity
            .cors(Customizer.withDefaults())
            .authorizeExchange { auth ->
                auth.anyExchange().authenticated()
            }
            .oauth2ResourceServer { oauth2 ->
                oauth2.opaqueToken {
                    it.introspector(someIntrospector)
                }
            }
            .csrf { csrf -> csrf.disable() }
            .logout { logout -> logout.disable() }
            .build()
    }
```

### PR DESCRIPTION
**Pull request by dukbong**
### Original Code:
The getAuthorizedClients method retrieves a map of OAuth2 authorized clients from a WebSession. In the original code, if the session is null, the method assigns a default empty HashMap. This code does not explicitly assert that WebSession cannot be null, potentially leading to silent failures or unexpected behavior.

Closes #14975

### Suggested Refactoring:
To ensure robustness and improve error handling, you can refactor the method by adding an assertion to prevent a null session. This approach uses Assert.notNull to raise an exception with a clear error message if the WebSession is null, helping developers quickly identify and fix the problem.

### CODE REVIEW

**Comment by dukbong** [comment]
> Hi @dukbong, thanks for the PR!
> 
> 
> 
> Just a heads up, I closed the original issue with [this explanation](https://github.com/spring-projects/spring-security/issues/14975#issuecomment-2090702647). I appreciate what you're doing in this change by asserting the session cannot be null. I'm wondering though if this change is absolutely necessary?
> 
> 
> 
> > This code does not explicitly assert that WebSession cannot be null, potentially leading to silent failures or unexpected behavior.
> 
> 
> 
> Can you explain the situation in which this would occur (other than the invalid configuration referred to in the original issue)?


Hi @sjohnr 

Initially, I planned to submit a pull request to address the issue, but as I reviewed the code, I found the structure to be questionable, prompting a change in direction.

The `getAuthorizedClient` function is written with the assumption that the session might be null. However, if the session is null, this leads to an exception in the `saveAuthorizedClient` function. To prevent this, I believe it's better to ensure the session isn't null either in the `getAuthorizedClient` function or in the `doOnSuccess` block.

Based on these observations, I created a pull request to address the issue.

- Your feedback is invaluable for my growth and learning.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java [hunk 1]
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
...
-		if (authorizedClients == null) {
-			authorizedClients = new HashMap<>();
-		}
-		return authorizedClients;
+		Assert.notNull(session, "WebSession cannot be null. Ensure the session is properly initialized before accessing authorized clients.");
+	    Map<String, OAuth2AuthorizedClient> authorizedClients = session.getAttribute(this.sessionAttributeName);
+	    return authorizedClients != null ? authorizedClients : new HashMap<>();
 	}
 
+
```
Comment (sjohnr) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:103]: Please remove this extra newline.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java
Comment (sjohnr) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:1]: Please update the copyright year in the header as per [contributing guidelines](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc#submit-a-pull-request).

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java [hunk 2]
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
...
 	private Map<String, OAuth2AuthorizedClient> getAuthorizedClients(WebSession session) {
-		Map<String, OAuth2AuthorizedClient> authorizedClients = (session != null)
-				? (Map<String, OAuth2AuthorizedClient>) session.getAttribute(this.sessionAttributeName) : null;
-		if (authorizedClients == null) {
-			authorizedClients = new HashMap<>();
-		}
-		return authorizedClients;
+		Assert.notNull(session, "WebSession cannot be null. Ensure the session is properly initialized before accessing authorized clients.");
+	    Map<String, OAuth2AuthorizedClient> authorizedClients = session.getAttribute(this.sessionAttributeName);
+	    return authorizedClients != null ? authorizedClients : new HashMap<>();
```
Comment (sjohnr) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:100]: You can add parentheses around the ternary condition to satisfy checkstyle and allow the build to pass.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java [hunk 3]
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
...
 
-	@SuppressWarnings("unchecked")
 	private Map<String, OAuth2AuthorizedClient> getAuthorizedClients(WebSession session) {
-		Map<String, OAuth2AuthorizedClient> authorizedClients = (session != null)
-				? (Map<String, OAuth2AuthorizedClient>) session.getAttribute(this.sessionAttributeName) : null;
-		if (authorizedClients == null) {
-			authorizedClients = new HashMap<>();
-		}
-		return authorizedClients;
+		Assert.notNull(session, "WebSession cannot be null. Ensure the session is properly initialized before accessing authorized clients.");
```
Comment (sjohnr) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:98]: I think the message can simply be `session cannot be null`.

**Review by sjohnr — CHANGES_REQUESTED** [review]
Thanks for the PR @dukbong. In addition to the feedback inline below, can you please add a test that asserts an `IllegalArgumentException` is thrown when `WebSession` is null?

**Comment by dukbong** [comment]
> Thanks for the PR @dukbong. In addition to the feedback inline below, can you please add a test that asserts an `IllegalArgumentException` is thrown when `WebSession` is null?

Thanks @sjohnr  for agreeing with my feedback.
I will make the changes and additions as per the comments below!

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java [hunk 4]
```diff
@@ -94,14 +94,10 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
 		// @formatter:on
 	}
 
-	@SuppressWarnings("unchecked")
```
Comment (sjohnr) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:97]: I don't think this should be removed. Can you restore this line?

**Review by sjohnr — COMMENTED** [review]
Thanks @dukbong. I added one minor comment below. Can you also please rebase your changes on main and squash commits? I will schedule this to be merged in the next minor so you won't see it merged for a few weeks.

**Comment by dukbong** [comment]
> Thanks @dukbong. I added one minor comment below. Can you also please rebase your changes on main and squash commits? I will schedule this to be merged in the next minor so you won't see it merged for a few weeks.


Hello. @sjohnr 

As you suggested, I rebased onto the main branch and squashed the commits into one.

If there's anything I did wrong or need to correct, please let me know!
I'll fix it right away!!

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java (hunk 4 above)
Comment (sjohnr) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:97]: Apologies as you were correct to remove this. I'll re-remove it in a polish commit.

**Comment by sjohnr** [comment]
Thanks @dukbong, this is now merged into `5.8.x` as 1695d03b72d9d48f24abd8381f5f120b99a7f7a1. I did apply a polish commit with some adjustments because I didn't run the build to see checkstyle and other build output, so check out 
e34621ec2ced820accc121b053752387d053d9a5 and let me know if you have questions.
